black.py

```diff
@@ -64,28 +66,38 @@ class CannotSplit(Exception):
 @click.argument(
     'src',
     nargs=-1,
-    type=click.Path(exists=True, file_okay=True, dir_okay=True, readable=True),
+    type=click.Path(
+        exists=True, file_okay=True, dir_okay=True, readable=True, allow_dash=True
+    ),
 )
 @click.pass_context
 def main(ctx: click.Context, line_length: int, fast: bool, src: List[str]) -> None:
     """The uncompromising code formatter."""
     sources: List[Path] = []
-    for s in src:
-        p = Path(s)
-        if p.is_dir():
-            sources.extend(gen_python_files_in_dir(p))
-        elif p.is_file():
-            # if a file was explicitly given, we don't care about its extension
-            sources.append(p)
-        else:
-            err(f'invalid path: {s}')
+    if len(src) == 1 and src[0] == '-':
+        sources.append(Path('-'))
+    else:
+        for s in src:
+            p = Path(s)
+            if p.is_dir():
+                sources.extend(gen_python_files_in_dir(p))
+            elif p.is_file():
+                # if a file was explicitly given, we don't care about its extension
+                sources.append(p)
+            elif s == '-':
+                err('Cannot format both stdin and other paths')
+            else:
+                err(f'invalid path: {s}')
     if len(sources) == 0:
         ctx.exit(0)
     elif len(sources) == 1:
         p = sources[0]
         report = Report()
         try:
-            changed = format_file_in_place(p, line_length=line_length, fast=fast)
+            if str(p) == '-':
```

Comment:
Since `-` is an unlikely real file name, it would be better to say `if not p.is_file() and str(p) == '-':`